invokeai/app/services/shared/invocation_context.py

```diff
@@ -434,6 +437,55 @@ def search_by_attrs(
             model_format=format,
         )
 
+    def download_and_cache_model(
+        self,
+        source: str | AnyHttpUrl,
+    ) -> Path:
+        """
+        Download the model file located at source to the models cache and return its Path.
+
+        This can be used to single-file install models and other resources of arbitrary types
+        which should not get registered with the database. If the model is already
+        installed, the cached path will be returned. Otherwise it will be downloaded.
+
+        Args:
+            source: A model path, URL or repo_id.
+
+        Returns:
+            Path to the downloaded model
+        """
+
+        return self._services.model_manager.install.download_and_cache_model(source=source)
+
+    def load_and_cache_model(
+        self,
+        source: Path | str | AnyHttpUrl,
+        loader: Optional[Callable[[Path], dict[str, Tensor]]] = None,
+    ) -> LoadedModelWithoutConfig:
+        """
+        Download, cache, and load the model file located at the indicated URL.
+
+        If the model is already downloaded, it will be loaded from the cache.
+
+        If the a loader callable is provided, it will be invoked to load the model. Otherwise,
+        `safetensors.torch.load_file()` or `torch.load()` will be called to load the model.
+
+        Be aware that the LoadedModelWithoutConfig object has no `config` attribute
+
+        Args:
+            source: A model Path, URL, or repoid.
+            loader: A Callable that expects a Path and returns a dict[str|int, Any]
+
+        Returns:
+            A LoadedModelWithoutConfig object.
+        """
+
+        if isinstance(source, Path):
+            return self._services.model_manager.load.load_model_from_path(model_path=source, loader=loader)
+        else:
+            model_path = self._services.model_manager.install.download_and_cache_model(source=str(source))
+            return self._services.model_manager.load.load_model_from_path(model_path=model_path, loader=loader)
```

Comment:
I don't love that we switch behaviour based on whether `source` is a `Path` or a `str`. It feels like a fragile distinction, especially given the popularity of using `str`s to represent paths in Python. The caller should always know whether they are dealing with a path or a URL/repo name, so I think it's better to make this distinction explicit.

In [this discussion](https://github.com/invoke-ai/InvokeAI/pull/6132#discussion_r1576894928) we had landed on an API that didn't require this type condition. Was there a reason for moving away from that?

Reply:
It looks like @psychedelicious removed `load_model_from_url()` in commit b12444 . Add it back?

Reply:
I moved the method from the high-level `services.model_manager` class to the invocation context. The function is otherwise the same.

Reply:
@psychedelicious  I can’t find `load_model_from_url()` in `invocation_context.py`, and the commit indicates that it was deleted from the model manager and then the `load_and_cache_model()` was modified to no longer call it.

I’ll take care of fixing the API if the change wasn’t deliberate.

Reply:
For clarity, I have renamed the methods `load_local_model()` and `load_remote_model()`. The former accepts the Path to a file or directory, and the latter accepts either a direct download URL or a HuggingFace URL. I have fixed the documentation and updated the pull request description.

Reply:
Sounds good!